### ISSUE TITLE
"dev" branch logical operations example FYI

### ISSUE THREAD

**Issue by shleym2000** [Artelnics/opennn#102]
Good news: now logical operations example can be executed without crashes or exceptions
Bad news: Results of that execution do not look like training was successful:

> OpenNN. Logical Operations Example.
> Setting data file preview...
> Setting rows labels...
> Setting columns names...
> Setting columns types...
> Setting data dimensions...
> Reading data...
> Data read succesfully...
> Checking binary columns...
> Training with quasi-Newton method...
> Epoch 1: Minimum parameters increment norm reached.
> Parameters increment norm: 0
> Parameters norm: 4.31773
> Training error: 1.40182
> Gradient norm: 0.0955539
> Training rate: 0
> Elapsed time: 00:00:00
> history: -1.40383e-17
> X Y AND OR NAND NOR XOR XNOR
> 
> X = 1 Y = 1
> 1 1  0.148278  0.168519  0.122213  0.400842 0.0798118 0.0803364
> 
> X = 1 Y = 0
> 1 0  0.151575  0.139584  0.146033  0.369082 0.0966274 0.0970989
> 
> X = 0 Y = 1
> 0 1  0.139005  0.141456   0.12921  0.411975 0.0789831 0.0993712
> 
> X = 0 Y = 0
> 0 0 0.135127 0.121931 0.142011 0.373413 0.108238  0.11928
> 
> C:\MyProjects\x64\Release\NeuralNetwork.exe (process 12704) exited with code 0.
>

**Comment by shleym2000** [comment]
Problem is in the fact that very first training iteration was considered being enough:
**_Epoch 1: Minimum parameters increment norm reached._**

It happened in the following code in quasi_newton_method.cpp file

>         if(optimization_data.**parameters_increment_norm** <= **minimum_parameters_increment_norm**)
>         {
>             if(display)
>             {
>                cout << "Epoch " << epoch << ": Minimum parameters increment norm reached.\n"
>                     << "Parameters increment norm: " << optimization_data.parameters_increment_norm << endl;
>             }
> 

The optimization_data.**parameters_increment_norm** is getting updated in the quasi_newton_method.h file here:

>        optimization_data.parameters_increment = optimization_data.training_direction*optimization_data.learning_rate;
> 
>        optimization_data.old_parameters = optimization_data.parameters;
> 
>        optimization_data.parameters += optimization_data.parameters_increment;
> 
>        optimization_data.**parameters_increment_norm** = _l2_norm_(optimization_data.parameters_increment);
> 
>        optimization_data.old_gradient = back_propagation.gradient;
> 
>        optimization_data.old_inverse_hessian = optimization_data.inverse_hessian;
> 
>        optimization_data.old_learning_rate = optimization_data.learning_rate;
>

**Comment by shleym2000** [comment]
As of now, training for logical operations test still ends after the epoch #0 when it is obviously not sufficient:

_Training with quasi-Newton method...
Epoch 0: **Minimum parameters increment norm reached**.
Parameters increment norm: 0
Parameters norm: 4.88395
Training error: 1.37488
Gradient norm: 0.103023_

**Comment by shleym2000** [comment]
After recent changes to the OpenNN **dev** library code the _logical operations_ example is getting trained for 16 iterations, but the results after training are still bogus:

_...
Epoch 16: Minimum loss decrease (0) reached.
Loss decrease: 0.00115466
Parameters norm: 6.53951
Training error: 1.36004
Gradient norm: 0.251057
Training rate: 0.962231
Elapsed time: 00:00:01
X Y AND OR NAND NOR XOR XNOR
X = 1 Y = 1
1 1 0.000372729    0.998939 9.01688e-05 3.33385e-05 0.000544097 2.10599e-05
X = 1 Y = 0
1 0 0.000560346     0.99839 0.000147943 5.90686e-05 0.000804925 3.77794e-05
X = 0 Y = 1
0 1 0.000546808     0.99845 0.000135257 5.33668e-05 0.000779516  3.5164e-05
X = 0 Y = 0
0 0  0.00221661    0.993399 0.000705854 0.000354622  0.00307801 0.000246252_

**Comment by shleym2000** [comment]
Yesterday test trained for 16 iterations, but produced incorrect results.
After last 24 hours changes test again trains only for 1 iteration. Results are still wrong.

**Comment by shleym2000** [comment]
I would blame recent changes to **learning_rate_algorithm.cpp** file ;)

**Comment by shleym2000** [comment]
Current status of this test is crash:

> Training with quasi-Newton method...
> Assertion failed: dimensions_match(m_leftImpl.dimensions(), m_rightImpl.dimensions()), file C:\MyProjects\RTS\Trial\TestLib\SandBox\OpenNN\eigen\unsupported\Eigen\CXX11\src\Tensor\TensorAssign.h, line 122

The call sack at the time exception is thrown looks like:
> Eigen::TensorEvaluator<Eigen::TensorAssignOp<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> const ,Eigen::ThreadPoolDevice>::evalSubExprsIfNeeded(float * __formal) Line 122	C++
> Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> const ,Eigen::ThreadPoolDevice,1>::run(const Eigen::TensorAssignOp<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> & expr, const Eigen::ThreadPoolDevice & device) Line 149	C++
> Eigen::TensorDevice<Eigen::Tensor<float,3,0,__int64>,Eigen::ThreadPoolDevice>::operator=<Eigen::Tensor<float,3,0,__int64>>(const Eigen::Tensor<float,3,0,__int64> & other) Line 36	C++
> OpenNN::Layer::softmax_derivatives(const Eigen::Tensor<float,2,0,__int64> & combinations, Eigen::Tensor<float,2,0,__int64> & activations, Eigen::Tensor<float,3,0,__int64> & activations_derivatives) Line 1566	C++
> OpenNN::ProbabilisticLayer::calculate_activations_derivatives(const Eigen::Tensor<float,2,0,__int64> & combinations_2d, Eigen::Tensor<float,2,0,__int64> & activations, Eigen::Tensor<float,3,0,__int64> & activations_derivatives) Line 282	C++
> OpenNN::ProbabilisticLayer::forward_propagate(const Eigen::Tensor<float,2,0,__int64> & inputs, OpenNN::Layer::ForwardPropagation & forward_propagation) Line 192	C++
> OpenNN::NeuralNetwork::forward_propagate(const OpenNN::DataSet::Batch & batch, OpenNN::NeuralNetwork::ForwardPropagation & forward_propagation) Line 361	C++
> OpenNN::QuasiNewtonMethod::perform_training() Line 1236	C++
> OpenNN::TrainingStrategy::perform_training() Line 1070	C++
>

**Comment by GuillermoArtelnics** [comment]
Hello shleym,

Yes, right now we are having a lot of changes for a better performance on dev, at the moment the regression problems are working well, now we are fine tuning the classification problems.

Thank you very much for your contributions to this project.

**Comment by shleym2000** [comment]
This test still shows exception and here is additional information you might find helpful in fixing this problem.

Exception is thrown from the following line:

>    void calculate_hidden_delta_probabilistic(Layer* next_layer_pointer,
>                                              Tensor<type, 3>& activations_derivatives_3d,
>                                              const Tensor<type, 2>& next_layer_delta,
>                                              Tensor<type, 2>& hidden_delta) const
>    {
> ...
>             case Device::EigenThreadPool:
>             {
>                ThreadPoolDevice* thread_pool_device = device_pointer->get_eigen_thread_pool_device();
> 
>                hidden_delta.device(*thread_pool_device) = next_layer_delta.contract(next_synaptic_weights, A_BT);
> 
>                **_hidden_delta.device(*thread_pool_device) = hidden_delta*activations_derivatives;_**
> 
>                return;
>             }
>        }
>    }
> 

The problem is in the difference in dimensions of the tensors on the right and on the left of the assignment operator. They must be the same for tensor assignment to work here.

**hidden_delta.device(*thread_pool_device)** has dimensions {4, 6}
**hidden_delta*activations_derivatives** product has dimensions {0, 6}

Checking dimensions of the right and left components of the successful tensor multiplication I see
 **hidden_delta** product has dimensions {4, 6}
**activations_derivatives** product has dimensions {0, 6}

**Comment by shleym2000** [comment]
Code changed again and I had to switch test to use **device(OpenNN::Device::EigenDefault)**.

I also added diagnostics to **case Device::EigenDefault:** logic of **calculate_hidden_delta_probabilistic** - same as you added to **case Device::EigenThreadPool:**

```
            case Device::EigenDefault:
            {
                DefaultDevice* default_device = device_pointer->get_eigen_default_device();
                hidden_delta.device(*default_device) = next_layer_delta.contract(next_synaptic_weights, A_BT);
                cout << "Hidden delta 1: " << hidden_delta << endl;
                cout << "Activations derivatives: " << activations_derivatives << endl;
                hidden_delta.device(*default_device) = hidden_delta*activations_derivatives;
                cout << "Hidden delta 2: " << hidden_delta << endl;
                return;
            }
```
Output at this point is: 
> Hidden delta 1: 5.82877e+17 7.84184e+17 4.49478e+17 5.94794e+17 5.70418e+17 5.72814e+17                                                                                                                 5.82877e+17 7.84184e+17 4.49478e+17 5.94794e+17 5.70418e+17 5.72814e+17                                                                                                                                 5.82877e+17 7.84184e+17 4.49478e+17 5.94794e+17 5.70418e+17 5.72814e+17                                                                                                                                 5.82877e+17 7.84184e+17 4.49478e+17 5.94794e+17 5.70418e+17 5.72814e+17
>Activations derivatives:
>Assertion failed: dimensions_match(m_leftImpl.dimensions(), m_rightImpl.dimensions()) ...

Same mismatch problem
**hidden_delta.device**(*thread_pool_device) has dimensions {4, 6}
**hidden_delta * activations_derivatives** product has dimensions {0, 6}

The possible problem could be when **activations_derivatives** gets defined:
`TensorMap < Tensor<type, 2> > activations_derivatives(activations_derivatives_3d.data(), batch_instances_number, neurons_number);`

batch_instances_number here is = 0, neurons_number = 6. The result is tensor with dimentions {0, 6}

I wonder if tensor with first dimension 0 even reasonable. Judging by diagnostics output the first dimention 0 made the following dimension irrelevant and entire tensor effectively became of 0 size.

**Comment by shleym2000** [comment]
As of May 19, 2020 the Logical Operation example is crashing during the very first training iteration when attempting to assign tensors with mismatched dimensions

Tensor with dimensions { 6,6,1 } can not be assigned to Tensor with dimensions { 4, 6, 6 } 

The call stack at the point of crash:
> OpenNN_LogicalOperations.exe!Eigen::**TensorEvaluator<Eigen::TensorAssignOp**<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> const ,Eigen::ThreadPoolDevice>::evalSubExprsIfNeeded(float * __formal) Line 122

> OpenNN_LogicalOperations.exe!Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> const ,Eigen::ThreadPoolDevice,1>::run(const Eigen::TensorAssignOp<Eigen::Tensor<float,3,0,__int64>,Eigen::Tensor<float,3,0,__int64> const> & expr, const Eigen::ThreadPoolDevice & device) Line 149

> OpenNN_LogicalOperations.exe!Eigen::TensorDevice<Eigen::Tensor<float,3,0,__int64>,Eigen::ThreadPoolDevice>::**operator=**<Eigen::Tensor<float,3,0,__int64>>(const Eigen::Tensor<float,3,0,__int64> & other) Line 36

> OpenNN_LogicalOperations.exe!OpenNN::**Layer::softmax_derivatives**(const Eigen::Tensor<float,2,0,__int64> & combinations, Eigen::Tensor<float,2,0,__int64> & activations, Eigen::Tensor<float,3,0,__int64> & activations_derivatives) Line 1228

> OpenNN_LogicalOperations.exe!OpenNN::ProbabilisticLayer::**calculate_activations_derivatives**(const Eigen::Tensor<float,2,0,__int64> & combinations_2d, Eigen::Tensor<float,2,0,__int64> & activations, Eigen::Tensor<float,3,0,__int64> & activations_derivatives) Line 727

> OpenNN_LogicalOperations.exe!OpenNN::ProbabilisticLayer::**forward_propagate**(const Eigen::Tensor<float,2,0,__int64> & inputs, OpenNN::Layer::ForwardPropagation & forward_propagation) Line 163

> OpenNN_LogicalOperations.exe!OpenNN::NeuralNetwork::**forward_propagate**(const OpenNN::DataSet::Batch & batch, OpenNN::NeuralNetwork::ForwardPropagation & forward_propagation) Line 1352

> OpenNN_LogicalOperations.exe!OpenNN::QuasiNewtonMethod::**perform_training**() Line 1126

> OpenNN_LogicalOperations.exe!OpenNN::TrainingStrategy::**perform_training**() Line 1070

**Comment by shleym2000** [comment]
FYI The current status of this example is bellow.
Results are wrong for the obvious reason - there was only one training iteration.
And the there was only one iteration because
"_Minimum parameters increment norm reached.
Parameters increment norm: 0_"
There could be no increment if this is the very first iteration and there is nothing to compare it with.
I think either there is no point to check this criteria in epoch == 0 or parameter_increment_norm value should be "faked" to high number for epoch == 0
```
OpenNN. Logical Operations Example.
Setting data file preview...
Setting rows labels...
Setting columns names...
Setting columns types...
Setting data dimensions...
Reading data...
Data read succesfully...
Checking binary columns...
Training with quasi-Newton method...
OpenNN Warning: Gradient norm is inf.
0
Epoch 1: OpenNN. Logical Operations Example.
Setting data file preview...
Setting rows labels...
Setting columns names...
Setting columns types...
Setting data dimensions...
Reading data...
Data read succesfully...
Checking binary columns...
Training with quasi-Newton method...
OpenNN Warning: Gradient norm is inf.
0
Epoch 1: Minimum parameters increment norm reached.
Parameters increment norm: 0
Parameters norm: 7.81553
Training error: -nan(ind)
Gradient norm: inf
Learning rate: 0
Elapsed time: 00:00:00
X Y AND OR NAND NOR XOR XNOR
X = 1 Y = 1
1 1   0.156958   0.122543   0.116213 0.00651151   0.573664  0.0241112
X = 1 Y = 0
1 0  0.209763   0.02324 0.0694337 0.0208716  0.007077  0.669615
X = 0 Y = 1
0 1  0.157419  0.117987  0.277483 0.0185979  0.247421  0.181092
X = 0 Y = 0
0 0  0.0702286  0.0307842  0.0529099  0.0286193 0.00213591   0.815322

...\x64\Debug\OpenNN_LogicalOperations.exe (process 12960) exited with code 0.
```

**Comment by shleym2000** [comment]
It looks like you made this simple example working again.

The current results bellow can be seen correctly is we would treat values larger than 0.3 as 1 and smaller than 0.02 as 0

```
...
Epoch 192: Minimum parameters increment norm reached.
Parameters increment norm: 0
Parameters norm: 7.63786
Training error: 0.696064
Gradient norm: 2.23998e-09
Learning rate: 0
Elapsed time: 00:00:04
X Y AND OR NAND NOR XOR XNOR
X = 1 Y = 1
1 1    0.480029      0.4951  0.00542991 7.31613e-05   0.0192939 7.48495e-05
X = 1 Y = 0
1 0 0.00847967   0.336458    0.33079 0.00442591   0.315375 0.00447116
X = 0 Y = 1
0 1 0.00847967   0.336458    0.33079 0.00442591   0.315375 0.00447116
X = 0 Y = 0
0 0 5.33778e-05  0.00262698    0.336453    0.323794   0.0134415    0.323631
```
